### PR TITLE
Add build definition to test TypeScript compatibility of office-ui-fabric-react

### PR DESCRIPTION
**Pull request by KevinTCoughlin**
### Overview

This PR adds a build definition that tests the compilation of `office-ui-fabric-react` against a range of TypeScript versions contained in a repository I will make public once this is merged.

Initially, the build trigger is pinned to a specific branch for testing end-to-end and to avoid resource contention.

I will follow-up with an iteration that modifies the triggers to contain `office-ui-fabric-react` tagged releases.

### Why?

The goal is to be notified when an incompatible typing is introduced so that we can:

1. Make a decision on whether or not to break compatibility for a given version
1. Provide a resource for partners to be aware of our TypeScript compatibility across major and minor releases of TypeScript.

---

### Build Overview

This build definition consists of two jobs with their runtime cost (tested on free DevOps agents) described below:

#### Job 1 - Build & Pack OUFR (~8 mins)

1. The first job clones this repository.
1. The job then installs dependencies and builds `office-ui-fabric-react` directly.
1. Finally it runs `npm pack` to produce the `office-ui-fabric-react` tarball and publishes it as an artifact for the second job to consume.

**Note:** The need for this job can likely be eliminated entirely in a subsequent iteration.

#### Job 2 - Compile OUFR in Sample App (~1 min per)

1. The second job depends on the first job _succeeding_ as it requires the `office-ui-fabric-react` tarball.
1. It is a matrix job that executes a maximum of two jobs in parallel which test compatibility for a given TypeScript version.
1. Each job clones the repository containing our sample applications consuming `office-ui-fabric-react`
1. Then the `office-ui-fabric-react` tarball is installed to the project. This is done to ensure we are not consuming a different release via NPM.
1. The job finally runs `tsc` in the root of the sample application and reports success or failure as build status.

**Note:** 30 seconds of the above runtime is the compliance step 😄.

#### Example Build Output

image

---

**Related Issues:** https://github.com/OfficeDev/office-ui-fabric-react/issues/8458 https://github.com/OfficeDev/office-ui-fabric-react/issues/8051 https://github.com/OfficeDev/office-ui-fabric-react/issues/7054

cc: @aftab-hassan @Vitalius1 @natalieethell @patmill 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8961)

### CODE REVIEW

File: scripts/typescript-compatibility-build.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+      - task: NodeTool@0
+        displayName: 'Use Node $(NodeVersion)'
+        inputs:
+          versionSpec: '$(NodeVersion)'
+          checkLatest: true
+
+      # Install Rush repo dependencies
+      - script: |
+          node common/scripts/install-run-rush.js install --bypass-policy
```
Comment (ecraig12345) [inline scripts/typescript-compatibility-build.yml]: Love the granular steps with descriptive names!

File: scripts/typescript-compatibility-build.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  - job: Compatibility
+    dependsOn: Build
+    condition: succeeded()
+    timeoutInMinutes: 60
+    pool:
+      vmImage: $(VmImage)
+    strategy:
+      maxParallel: 2
+      matrix:
+        'TypeScript 2.4':
```
Comment (ecraig12345) [inline scripts/typescript-compatibility-build.yml]: Do we expect current OUFR to build against 2.4 and 2.7? Especially for 2.4, I'd expect a huge number of breaks.

File: scripts/typescript-compatibility-build.yml (hunk 1 above)
Comment (KevinTCoughlin) [reply scripts/typescript-compatibility-build.yml]: Thank you! 🙇‍♂️

File: scripts/typescript-compatibility-build.yml (hunk 2 above)
Comment (KevinTCoughlin) [reply scripts/typescript-compatibility-build.yml]: Correct, OUFR >= `6.x` does not build against `2.4`. @aditima and I think that this is useful information to surface as part of the build. One of the friction points today is that SPFx developers use various versions of the compiler, `2.4` being one of them ([rush-stack-compiler releases](https://github.com/microsoft/web-build-tools/tree/master/stack)).

We want to be able to direct folks to this running build as a way for them to decide what version of TypeScript they need to upgrade to in order to be compatible. @aditima had a good point that if we only surface passing builds then folks would still be unsure what versions may lead to problems.

We'll figure out how to handle the build alerting spam that comes along with this decision as part of wiring that up.

**Pull request merged**